### ISSUE TITLE
Config sync was broken on both masters (2.11.10)

### ISSUE THREAD

**Issue by K0nne** [Icinga/icinga2#8938]
## Describe the bug

Today we were investigating a bug with some new installed icinga-agents which have a long term flapping-problem.
(hosts are some hours available, then some hours not). we discovered that the config sync on both masters is broken, because the timestamp on` /var/lib/icinga2/api/zones` differs greatly on both machines (current time on the config master vs. 20.7.)

The secondary master reported the problem correctly. unfortunately we did not saw the alert:

`Icinga 2 has been running for 37 minutes and 27 seconds. Version: v2.11.8-4-g57e8fd392; Last zone sync stage validation failed at 2021-07-20 13:04:10 +0200` (at this time we were still at 2.11.8.4)

I know of a config problem, which I had fixed at this time. It was fixed at the config master (primary master) and after everything was fine, I made a config check, did a reload on the config master and was expecting everything was fine again. At this time the config sync was broken. 

I'm not sure if I am expecting wrong things, but I thought the fixed config whould be synced to the secondary master, the content of zones-stage whould be overwritten, then synced and everything is fine again. Do I have to fix config problems on both masters?

When I look at the secondary master now, zones-stage seems to be in a sync-loop. the config comes and goes.
The file `status `of the old config contains "1" and `startup.log` still points at the config problem, which was fixed on 20.7.

The config problem from 20.7. was as following:

A host, which was created via api was referencing a hostgroup, wich was created some time before in the filesystem. I fixed this by deleting the hostgroup from the filesystem and re-creating the hostgroup via api.

Maybe this has broken the sync somehow.

To fix this problem we stopped the second master and removed all content of /var/lib/icinga2/api/ and started again. Now both zones directories are in sync again.

## Expected behavior

The config sync should be working.

## Screenshots

comparison: zones on both master systems
![27-07-_2021_11-19-53](https://user-images.githubusercontent.com/34264690/127144319-ef2bca41-6034-43cc-8696-8b47b5aae46b.png)

zones-stage on the secondary master
![27-07-_2021_11-21-51](https://user-images.githubusercontent.com/34264690/127144333-f4ad124a-cf19-480f-b34d-394b51c63362.png)

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`):
```
icinga2 - The Icinga 2 network monitoring daemon (version: 2.11.10-1)

Copyright (c) 2012-2021 Icinga GmbH (https://icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

System information:
  Platform: Red Hat Enterprise Linux Server
  Platform version: 7.9 (Maipo)
  Kernel: Linux
  Kernel version: 3.10.0-1160.31.1.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: runner-hh8q3bz2-project-505-concurrent-0

Application information:

General paths:
  Config directory: /etc/icinga2
  Data directory: /var/lib/icinga2
  Log directory: /var/log/icinga2
  Cache directory: /var/cache/icinga2
  Spool directory: /var/spool/icinga2
  Run directory: /run/icinga2

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid
```
* Operating System and version: 

`rhel 7.9`

* Enabled features (`icinga2 feature list`):
```
Disabled features: command compatlog debuglog elasticsearch gelf livestatus opentsdb perfdata statusdata syslog
Enabled features: api checker graphite ido-mysql influxdb mainlog notification
```
* Icinga Web 2 version and modules (System - About):
```
Icinga Web 2 Version
    2.7.5
Git Commit
    18996270b264976adf18d20da557d0c2806217c5
PHP Version
    7.1.8
Git Commit Datum
    2021-07-12
Copyright
    © 2013-2021 Das Icinga Projekt 
```
* Config validation (`icinga2 daemon -C`):
```
[2021-07-27 13:15:17 +0200] information/cli: Icinga application loader (version: 2.11.10-1)
[2021-07-27 13:15:17 +0200] information/cli: Loading configuration file(s).
[2021-07-27 13:15:20 +0200] information/ConfigItem: Committing config item(s).
[2021-07-27 13:15:20 +0200] information/ApiListener: My API identity: master01
[2021-07-27 13:15:30 +0200] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 56, rate: 84.9333/s (5096/min 5096/5min 5096/15min);
[2021-07-27 13:15:30 +0200] information/WorkQueue: #5 (GraphiteWriter, graphite) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-07-27 13:15:30 +0200] information/WorkQueue: #7 (InfluxdbWriter, influxdb-b) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-07-27 13:15:30 +0200] information/WorkQueue: #6 (InfluxdbWriter, influxdb-a) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-07-27 13:15:30 +0200] information/WorkQueue: #10 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-07-27 13:15:30 +0200] information/WorkQueue: #9 (ApiListener, RelayQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 GraphiteWriter.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 2 InfluxdbWriters.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 NotificationComponent.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 IdoMysqlConnection.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 CheckerComponent.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 User.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 6 TimePeriods.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 5956 Zones.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 116135 Services.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 65 ScheduledDowntimes.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 163804 Notifications.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 NotificationCommand.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 42 Comments.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 IcingaApplication.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 5961 Endpoints.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 203 HostGroups.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 23847 Hosts.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 511 Downtimes.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 362 CheckCommands.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 FileLogger.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 9 ApiUsers.
[2021-07-27 13:15:47 +0200] information/ConfigItem: Instantiated 1 ApiListener.
[2021-07-27 13:15:47 +0200] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2021-07-27 13:15:47 +0200] information/cli: Finished validating the configuration file(s).
```

## Additional context

Is it possible to check the timestamp of both configs in the "zones" directory with the api? I was not expecting the config sync to break at any time.

**Comment by julianbrost** [comment]
> and startup.log still points at the config problem, which was fixed on 20.7.

Can you share that error message? If I understand correctly from the other statement in your report, this might be an error due to a duplicate object? If so, it should give both conflicting config locations. What are these locations?

**Comment by K0nne** [comment]
We already purged the startup.log, but I found an older copy of the file in a backup:

```
[2021-07-26 20:25:42 +0200] information/cli: Icinga application loader (version: 2.11.10-1)
[2021-07-26 20:25:42 +0200] information/cli: Loading configuration file(s).
[2021-07-26 20:25:45 +0200] information/ConfigItem: Committing config item(s).
[2021-07-26 20:25:45 +0200] information/ApiListener: My API identity: dxzmicinga02
[2021-07-26 20:25:46 +0200] critical/config: Error: Validation failed for object '%affected_host_1%' of type 'Host'; Attribute 'groups': Object '%hostgroup_located_in_filesystem_1%' of type 'HostGroup' does not exist.
Location: in /var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_1%.conf: 5:2-5:26
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_1%.conf(3):
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_1%.conf(4):  address = "%address%"
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_1%.conf(5):  groups = [ "%hostgroup_located_in_filesystem_1%" ]
                                                                                                                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_1%.conf(6):  vars["http_json_guid"] = "true"
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_1%.conf(7):  vars["http_json_health_uri"] = "%uri%"

[2021-07-26 20:25:46 +0200] critical/config: Error: Validation failed for object '%affected_host_2%' of type 'Host'; Attribute 'groups': Object '%hostgroup_located_in_filesystem_2%' of type 'HostGroup' does not exist.
Location: in /var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_2%.conf: 5:2-5:28
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_2%.conf(3):
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_2%.conf(4):  address = "%address%"
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_2%.conf(5):  groups = [ "%hostgroup_located_in_filesystem_2%" ]
                                                                                                                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_2%.conf(6):  vars["http_json_guid"] = "true"
/var/lib/icinga2/api/packages/_api/2cd6848f-6d55-4420-9b81-7f13f3425189/conf.d/hosts/%affected_host_2%.conf(7):  vars["http_json_health_uri"] = "%uri%"

[and so on. all the same error]

[2021-07-26 20:25:46 +0200] critical/config: 17 errors
[2021-07-26 20:25:47 +0200] critical/cli: Config validation failed. Re-run with 'icinga2 daemon -C' after fixing the config.
```

This config problem comes from a time when we created some hosts and hostgroups in the filesystem. At a later time we moved the creation of those hosts to a selfservice, which utilizes the api. Here the users added such an old hostgroup. 

API Hosts cant reference filesystem hostgroups. When this happens to config check fails. 

The lazy way to live with this whould be to comment the groups attibute of the affected host on the master, make 2 reloads until the host is synced to the satellites, then un-comment the groups attribute of the affected host on the config master, make another reload and the api-host is member of the hostgroup on the satellite. 

I wanted to resolve this in a sustained way. So I removed the hostgroup from the filesystem, added it again with the api and un-commented the groups attribute on the affected hosts on the config master. This should be reproduceable.

**Comment by widhalmt** [comment]
ref/NC/729839

**Comment by K0nne** [comment]
The hostgroups in question were originally located at /etc/icinga2/zones.d/master/groups.conf

**Comment by Al2Klimov** [comment]
ref/IP/34703

**Comment by Al2Klimov** [comment]
> API Hosts cant reference filesystem hostgroups. When this happens the config check fails.

Created a host group in `zones.d` and a member via API. Works for me. Or do I oversee something which matters for reproduction?

**Comment by K0nne** [comment]
Sorry for the late reply. I was on vacation.

Today the sync problem resurfaced again. This time on a satellite. The Icinga check was hinting "Last zone sync stage validation failed". During analysis I found the following log message on the satellite, which is interesting. When I checked the file did not exist.

```
[2021-08-12 10:22:19 +0200] critical/ApiListener: Could not create object 'HOSTGROUP_NAME':
[2021-08-12 10:22:19 +0200] critical/ApiListener: Error: Object 'HOSTGROUP_NAME' of type 'HostGroup' re-defined: in /var/lib/icinga2/api/packages/_api/FQDN/conf.d/hostgroups/HOSTGROUP_NAME.conf: 1:0-1:33; previous definition: in /var/lib/icinga2/api/packages/_api/FQDN/conf.d/hostgroups/HOSTGROUP_NAME.conf: 1:0-1:33
Location: in /var/lib/icinga2/api/packages/_api/FQDN/conf.d/hostgroups/HOSTGROUP_NAME.conf: 1:0-1:33
```

From this point onwards the satellite was unable to get a new configuration. The sync between the masters was ok. On the ConfigMaster I was able to query the hostgroup with icinga2 object list --name "HOSTGROUP_NAME" without problems. 

In web2 the hostgroup was available and the hosts were checked as if everything whould be fine:

![16-08-_2021_08-55-52](https://user-images.githubusercontent.com/34264690/129582393-8e28440c-e6a3-429b-85da-2559f6151b91.png)

The hostgroup in question was defined as following:

![16-08-_2021_09-02-41](https://user-images.githubusercontent.com/34264690/129580654-463b0b6c-50ca-449c-a82f-78146f67907a.png)

It might be related to the global zone "global-groups", which we use for API-created hostgroups. It is defined on all satellites.

from zones.conf
```
object Zone "global-groups" {
        global = true
}
```

After this I tried to change the loglevel from warning to information. Upon restarting icinga2 on the satellite, the process was unable to start:

```
sudo systemctl start icinga2
Job for icinga2.service failed because the control process exited with error code. See "systemctl status icinga2.service" and "journalctl -xe" for details.
```

journalctl was listing config errors because of that missing hostgroup. The Beginning of the state file was also containing those errors. After this I renamed the dir "/var/lib/icinga2/api" on the satellite and did a reload. The problem was gone.

Tomorrow I will try to reproduce it.
